### PR TITLE
Dev batch data extraction

### PR DESCRIPTION
**Pull request by alexisszmundy**
Allows BIG HISTO load test scenario
Extraction of large data from the database

### CODE REVIEW

File: src/main/java/fr/insee/genesis/domain/dtos/SurveyUnitUpdateDto.java [hunk 1]
```diff
@@ -19,4 +21,30 @@ public class SurveyUnitUpdateDto {
...
+        jsonObject.put("idQuest",idQuest);
+        jsonObject.put("idCampaign",idCampaign);
+        jsonObject.put("idUE",idUE);
+        jsonObject.put("state",state.toString());
+        jsonObject.put("mode",mode.getModeName());
+        jsonObject.put("recordDate",recordDate.toString());
+        jsonObject.put("fileDate",fileDate.toString());
+
+        JSONArray collectedVariablesJSONArray = new JSONArray();
+        if(collectedVariables != null)
```
Comment (alicela) [inline src/main/java/fr/insee/genesis/domain/dtos/SurveyUnitUpdateDto.java:36]: add { }

File: src/main/java/fr/insee/genesis/controller/rest/ResponseController.java [hunk 2]
```diff
@@ -217,11 +219,42 @@ public ResponseEntity<List<SurveyUnitUpdateDto>> findResponsesByUEAndQuestionnai
...
+
+        String filepathString = String.format("OUT/%s/OUT_ALL_%s.json", idQuestionnaire, LocalDateTime.now().toString().replace(":",""));
+        Path filepath = Path.of(fileUtils.getDataFolderSource(), filepathString);
+
+        if(surveyUnitDtos.size() <= Constants.BATCH_SIZE){
+            List<SurveyUnitUpdateDto> responses = surveyUnitService.findByIdQuestionnaire(idQuestionnaire);
+            JSONArray jsonResponses = JSONUtils.getJSONArrayFromResponses(responses);
+            fileUtils.appendJSONFile(filepath, jsonResponses);
+        }else{
+            log.info("More than " + Constants.BATCH_SIZE + " responses ! Using sequential extraction method to save RAM...");
```
Comment (alicela) [inline src/main/java/fr/insee/genesis/controller/rest/ResponseController.java:238]: log.info(" {}",)

File: src/main/java/fr/insee/genesis/domain/dtos/SurveyUnitDto.java [hunk 3]
```diff
@@ -15,4 +17,16 @@ public class SurveyUnitDto {
 
 	private Mode mode;
 
+	@Override
+	public boolean equals(Object o) {
+		if (this == o) return true;
```
Comment (alicela) [inline src/main/java/fr/insee/genesis/domain/dtos/SurveyUnitDto.java:22]: {}

File: src/main/java/fr/insee/genesis/domain/dtos/SurveyUnitDto.java [hunk 4]
```diff
@@ -15,4 +17,16 @@ public class SurveyUnitDto {
...
+		if (o == null || getClass() != o.getClass()) return false;
```
Comment (alicela) [inline src/main/java/fr/insee/genesis/domain/dtos/SurveyUnitDto.java:23]: {}

File: src/main/java/fr/insee/genesis/domain/dtos/SurveyUnitUpdateDto.java [hunk 5]
```diff
@@ -19,4 +21,30 @@ public class SurveyUnitUpdateDto {
...
+            for(CollectedVariableDto collectedVariableDto : collectedVariables)
+                collectedVariablesJSONArray.add(collectedVariableDto.toJSONObject());
```
Comment (alicela) [inline src/main/java/fr/insee/genesis/domain/dtos/SurveyUnitUpdateDto.java:38]: {}

File: src/main/java/fr/insee/genesis/domain/dtos/SurveyUnitUpdateDto.java [hunk 6]
```diff
@@ -19,4 +21,30 @@ public class SurveyUnitUpdateDto {
...
+
+        JSONArray externalVariablesJSONArray = new JSONArray();
+        if(externalVariables != null)
```
Comment (alicela) [inline src/main/java/fr/insee/genesis/domain/dtos/SurveyUnitUpdateDto.java:41]: {}

File: src/main/java/fr/insee/genesis/infrastructure/utils/FileUtils.java [hunk 7]
```diff
@@ -152,4 +158,58 @@ public String getSpecFolder(String campaign) {
...
+
+	/**
+	 * Write a text file.
+	 * @param filePath Path to the file.
+	 * @param fileContent Content of the text file.
+	 */
+	public void writeFile(Path filePath, String fileContent) {
+		Path path = Path.of(dataFolderSource, filePath.toString());
+		boolean fileCreated = false;
+		File myFile = null;
```
Comment (alicela) [inline src/main/java/fr/insee/genesis/infrastructure/utils/FileUtils.java:170]: Should we close file ?

File: src/main/java/fr/insee/genesis/infrastructure/utils/FileUtils.java [hunk 8]
```diff
@@ -152,4 +158,58 @@ public String getSpecFolder(String campaign) {
...
+	 * @param jsonArray JSON objects to write
+	 */
+	public void appendJSONFile(Path filePath, JSONArray jsonArray) {
+		String content = jsonArray.toJSONString();
+        File myFile;
+		try {
+			Files.createDirectories(filePath.getParent());
+			myFile = filePath.toFile();
+
+			try (RandomAccessFile raf = new RandomAccessFile(myFile, "rw")) {
```
Comment (alicela) [inline src/main/java/fr/insee/genesis/infrastructure/utils/FileUtils.java:200]: try in try...

File: src/main/java/fr/insee/genesis/infrastructure/utils/FileUtils.java (hunk 7 above)
Comment (loichenninger) [reply src/main/java/fr/insee/genesis/infrastructure/utils/FileUtils.java:170]: The FileWriter is open in a try-with-resources so it closes after the execution of the block

File: src/main/java/fr/insee/genesis/infrastructure/utils/FileUtils.java (hunk 8 above)
Comment (loichenninger) [reply src/main/java/fr/insee/genesis/infrastructure/utils/FileUtils.java:200]: I chose to leave the code as it is, as the refactors added unnecessary complexity in my opinion

File: src/main/java/fr/insee/genesis/infrastructure/utils/FileUtils.java [hunk 9]
```diff
@@ -152,4 +158,58 @@ public String getSpecFolder(String campaign) {
...
+				if(myFile.length() == 0)
+					raf.write("[]".getBytes(StandardCharsets.UTF_8));
+
+				raf.seek(myFile.length()-1);
+
+				if(myFile.length() != 2)
```
[the remark this answers is not in the document]
Comment (loichenninger) [reply src/main/java/fr/insee/genesis/infrastructure/utils/FileUtils.java:206]: We want to add a comma only if the file already contains json objects. If the file only contains "[]" (length equals 2) we don't want to add a comma.

**Pull request merged**